src/engine/BoundManager.cpp

```diff
@@ -296,3 +304,213 @@ void BoundManager::registerRowBoundTightener( IRowBoundTightener *ptrRowBoundTig
+    }
+    return tightened;
+}
+
+bool BoundManager::tightenUpperBound( unsigned variable, double value, const SparseUnsortedList &row )
+{
+    bool tightened = setUpperBound( variable, value );
+
+    if ( tightened )
+    {
+        if ( shouldProduceProofs() )
+            _boundExplainer->updateBoundExplanationSparse( row, true, variable );
+
+        if ( _tableau != nullptr )
+            _tableau->updateVariableToComplyWithUpperBoundUpdate( variable, value );
+    }
+
+    return tightened;
+}
+
+void BoundManager::resetExplanation( const unsigned var, const bool isUpper ) const
+{
+    _boundExplainer->resetExplanation( var, isUpper );
+}
+
+void BoundManager::setExplanation( const Vector<double> &explanation, unsigned var, bool isUpper ) const
+{
+    ASSERT( explanation.size() == _boundExplainer->getNumberOfRows() || explanation.empty() );
+    _boundExplainer->setExplanation( explanation, var, isUpper );
+}
+
+BoundExplainer *BoundManager::getBoundExplainer() const
+{
+    return _boundExplainer;
+}
+
+void BoundManager::setBoundExplainerContent( BoundExplainer *boundsExplainer )
+{
+    *_boundExplainer = *boundsExplainer;
+}
+
+bool BoundManager::addLemmaExplanation( unsigned var, double value, BoundType affectedVarBound,
+                                        unsigned causingVar, BoundType causingVarBound,
+                                        PiecewiseLinearFunctionType constraintType )
+{
+    if ( !shouldProduceProofs() )
+        return false;
+
+    ASSERT( causingVar < _tableau->getN() && var < _tableau->getN() );
+
+    // Register new ground bound, update certificate, and reset explanation
+    Vector<double> explanation( 0 );
+    explainBound( causingVar, causingVarBound, explanation );
+
+    bool tightened = affectedVarBound == UPPER ? tightenUpperBound( var, value ) : tightenLowerBound( var, value );
+
+    if ( tightened )
+    {
+        std::shared_ptr<PLCExplanation> PLCExpl = std::make_shared<PLCExplanation>( causingVar, var, value, causingVarBound, affectedVarBound, explanation, constraintType );
+        // TODO uncomment when proof producing engine is integrated
+        // _engine->getUNSATCertificateCurrentPointer()->addPLCExplanation( PLCExpl );
+        // affectedVarBound == UPPER ? _engine->updateGroundUpperBound( var, value ) : _engine->updateGroundLowerBound( var, value );
+        resetExplanation( var, affectedVarBound );
+    }
+    return true;
+}
+
+void BoundManager::registerEngine( IEngine *engine )
+{
+    _engine = engine;
+}
+
+void BoundManager::initializeBoundExplainer( unsigned numberOfVariables, unsigned numberOfRows )
+{
+    // TODO uncomment when proof producing engine is integrated
+    // if ( _engine->shouldProduceProofs() )
+    _boundExplainer = new BoundExplainer( numberOfVariables, numberOfRows, _context );
+}
+
+int BoundManager::getInconsistentVariable() const
+{
+    if ( _consistentBounds )
+        return -1;
```

Comment:
Please use an enumed value instead of this nameless -1
It also makes more sense to return unsigned (and the sentinel can be MAX_UNSIGNED - 1)